torch/csrc/jit/passes/create_functional_graphs.cpp

```diff
@@ -0,0 +1,233 @@
+        functional_graph_nodes.emplace_back(functional_subgraph_node);
+        functional_subgraph_node =
+            graph_->createWithSubgraph(prim::FunctionalGraph)->insertAfter(n);
+        SubgraphUtils::mergeNodeIntoSubgraph(n, functional_subgraph_node);
+      }
+    }
+    functional_graph_nodes.emplace_back(functional_subgraph_node);
+
+    for (Node* functional_node : functional_graph_nodes) {
+      if (!inlineIfTooSmall(functional_node)) {
+        ConstantPooling(functional_node->g(attr::Subgraph));
+      }
+    }
+    return changed;
+  }
+
+  bool AnalyzeFunctionalSubset(Node* n) {
+    bool functional_node = true;
+
+    // Functional Graphs are not responsible for maintaining aliasing
+    // relationships. If an output of a functional graph escapes scope
+    // or is mutated then we might change semantics of the program if
+    // aliasing relationships are changed.
+    // For now, we don't allow any values which are mutated into the functional
+    // graph, and we don't allow any nodes which have outputs that escape scope.
+    // Possible Future Improvements:
+    // - allow inputs to have mutations so long as there are no mutations in the
+    // graph
+    // - allow functional graphs to have at most one value that can escape scope
+    // - allow outputs which alias the wildcard set but do not "re-escape"
+
+    for (Value* v : n->outputs()) {
+      bool has_writers = aliasDb_->hasWriters(v);
+      bool escapes_scope = aliasDb_->escapesScope(v);
+      if (has_writers) {
+        mutated_values_.insert(v);
+      }
+      functional_node = functional_node && !escapes_scope && !has_writers;
+    }
+
+    for (Block* block : n->blocks()) {
+      auto functional_block = AnalyzeFunctionalSubset(block);
+      functional_node = functional_node && functional_block;
+    }
+
+    auto inputs = n->inputs();
+    functional_node = functional_node &&
+        std::all_of(inputs.begin(), inputs.end(), [&](Value* v) {
+                        return !mutated_values_.count(v);
+                      });
+    if (functional_node) {
+      functional_nodes_.insert(n);
+    }
+    return functional_node;
+  }
+
+  void AnalyzeFunctionalSubset(at::ArrayRef<Block*> blocks) {
+    for (Block* block : blocks) {
+      AnalyzeFunctionalSubset(block);
+    }
+  }
+
+  bool AnalyzeFunctionalSubset(Block* block) {
+    bool is_functional_block = true;
+    // block inputs will not yet have been iterated through,
+    // so we need to add them to our set of mutated & escape values.
+    for (Value* v : block->inputs()) {
+      bool has_writers = aliasDb_->hasWriters(v);
+      if (has_writers) {
+        mutated_values_.insert(v);
+      }
+    }
+    for (Node* n : block->nodes()) {
+      bool functional = AnalyzeFunctionalSubset(n);
+      is_functional_block = is_functional_block && functional;
+    }
+    return is_functional_block;
+  }
+
+  std::unordered_set<Node*> functional_nodes_;
+  std::unordered_set<Value*> mutated_values_;
```

Comment:
This seems redundant, since it's basically a cache for a `hasWriters` check. Do things get a lot slower if we remove it? If we must have it, it feels like something better placed in `AliasDb`.

Reply:
i had to do this because you're mutating the graph during the pass, so you'll hit assertion errors if you remove a node's input and then ask if it has writers. will comemtn

Reply:
There's a comment already actually:
```
    // First, analyze the functional subset of the graph, and then create
    // functional graphs. The graph gets mutated when we create functional
    // subgraphs, invalidating the AliasDb, so we need to do our analysis
    // first.
```